mybuild_docs/sphinx/man/2.10.0/hssiloopback.8
.\" Man page generated from reStructuredText.
.
.TH "HSSILOOPBACK" "8" "Nov 22, 2023" "2.10.0" "OPAE"
.SH NAME
hssiloopback \- High Speed Serial Interface ethernet loopback
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
hssiloopback [\-h] [\-\-pcie\-address PCIE_ADDRESS, \-P PCIE_ADDRESS] \-\-loopback [{enable,disable}]
.ft P
.fi
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
The \fBhssiloopback\fP  tool enables and disable ethernet loopback.
.SS OPTIONAL ARGUMENTS
.sp
\fB\-h, \-\-help\fP
.sp
Prints usage information
.sp
\fB\-\-pcie\-address PCIE_ADDRESS, \-P PCIE_ADDRESS\fP
.sp
The PCIe address of the desired fpga  in ssss:bb:dd.f format. sbdf of device to program (e.g. 04:00.0 or 0000:04:00.0). Optional when one device in system.
.sp
\fB\-\-loopback [{enable,disable}]\fP
.sp
Ethernet enable or disable loopback.
.SH EXAMPLES
.sp
\fBhssiloopback \-\-pcie\-address  0000:04:00.0 \-\-loopback enable\fP
.sp
Enables ethernet loopback
.sp
\fBhssiloopback \-\-pcie\-address  0000:04:00.0 \-\-loopback disable\fP
.sp
Disable ethernet loopback
.SH REVISION HISTORY
.sp
| Date | Intel Acceleration Stack Version | Changes Made |
|:——|—————————\-|:————–|
|2018.05.21| DCP 1.1 Beta (works with Quartus Prime Pro 17.1.1) | Made the following changes: Expanded descriptions of \fBnlb_mode_3\fP and\fBdma_afu\fP tests.  Added a second example command. |
.SH AUTHOR
Intel DCG FPT SW
.SH COPYRIGHT
2017 Intel Corporation
.\" Generated by docutils manpage writer.
.


binaries/hssi/ethernet/hssiloopback.py
#! /usr/bin/env python3
# Copyright(c) 2021-2022, Intel Corporation
#
# Redistribution  and  use  in source  and  binary  forms,  with  or  without
# modification, are permitted provided that the following conditions are met:
#
# * Redistributions of  source code  must retain the  above copyright notice,
#  this list of conditions and the following disclaimer.
# * Redistributions in binary form must reproduce the above copyright notice,
#  this list of conditions and the following disclaimer in the documentation
#   and/or other materials provided with the distribution.
# * Neither the name  of Intel Corporation  nor the names of its contributors
#   may be used to  endorse or promote  products derived  from this  software
#   without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
# IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
# LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
# CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
# SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
# INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
# CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

import argparse
import sys
from ethernet.hssicommon import *


class FPGAHSSILPBK(HSSICOMMON):
    def __init__(self, args):
        self._loopback = args.loopback
        self._hssi_grps = args.hssi_grps
        self._pcie_address = args.pcie_address
        self._port = args.port
        self._channel = args.ncsi_ch_sel
        HSSICOMMON.__init__(self)

    def hssi_loopback_en(self):
        """
        clear ctl address and ctl sts CSR
        write 0x7/0x8 value ctl address
        write 0x2 value ctl sts csr
        poll for status
        clear ctl address and ctl sts CSR
        """
        self.open(self._hssi_grps[0][0])

        hssi_feature_list = hssi_feature(self.read32(0, self.hssi_csr.HSSI_FEATURE_LIST))
        if (self._port >= self.hssi_csr.HSSI_PORT_COUNT):
            print("Invalid Input port number")
            self.close()
            return False

        enable = self.register_field_get(hssi_feature_list.port_enable,
                                         self._port)
        if enable == 0:
            print("Input port is not enabled or active")
            self.close()
            return False

        ctl_addr = hssi_ctl_addr(0)
        if self._loopback == 'enable':
            ctl_addr.sal_cmd = HSSI_SAL_CMD.ENABLE_LOOPBACK.value
        else:
            ctl_addr.sal_cmd = HSSI_SAL_CMD.DISABLE_LOOPBACK.value

        if self._channel != None and self._channel <=2: # add limit that is 2
            ncsi_ch_sel_reg = self.hssi_csr.HSSI_NCSI_CH_SEL
            self.write64(0, ncsi_ch_sel_reg, self._channel)
            print("NCSI register offset has been set to", hex(self.read64(0, ncsi_ch_sel_reg)))

        # set port number
        ctl_addr.port_address = self._port

        cmd_sts = hssi_cmd_sts(0)
        cmd_sts.value = 0x2

        ret = self.clear_ctl_sts_reg(0)
        if not ret:
            print("Failed to clear HSSI CTL STS csr")
            self.close()
            return False

        self.write32(0, self.hssi_csr.HSSI_CTL_ADDRESS, ctl_addr.value)
        self.write32(0, self.hssi_csr.HSSI_CTL_STS, cmd_sts.value)

        if not self.read_poll_timeout(0,
                                      self.hssi_csr.HSSI_CTL_STS,
                                      0x2):
            print("HSSI ctl sts csr fails to update ACK")
            self.close()
            return False

        ret = self.clear_ctl_sts_reg(0)
        if not ret:
            print("Failed to clear HSSI CTL STS csr")
            self.close()
            return False

        self.close()
        return True

    def hssi_loopback_start(self):
        """
        print hssi info
        enable/disable hssi loopback
        """
        if not self.hssi_info(self._hssi_grps[0][0]):
            print("Failed to read hssi information")
            return False
        if not self.hssi_loopback_en():
            return False
        return True


def main():
    """
    parse input arguments pciaddress and mtu
    enum fpga pcie devices and find match
    enable/disable loopback
    """
    parser = argparse.ArgumentParser()

    pcieaddress_help = 'sbdf of device to program \
                        (e.g. 0000:04:00.0).' \
                       ' Optional when one device in system.'
    parser.add_argument('--pcie-address', '-P',
                        default=None, help=pcieaddress_help)

    parser.add_argument('--loopback',
                        choices=['enable', 'disable'], nargs='?',
                        default=None,
                        help='loopback enable')

    parser.add_argument('--port', type=int,
                        default=0,
                        help='hssi port number')

    parser.add_argument('--ncsi_ch_sel', type=int,
                        default=None,
                        help='hssi NCSI_CH_SEL; Value 2 to disable NCSI')

    args = parser.parse_args()

    print("args", args)

    if args.pcie_address and not verify_pcie_address(args.pcie_address.lower()):
        sys.exit(1)

    if args.loopback is None:
        print('please specify --loopback enable/disable')
        sys.exit(1)

    args.hssi_grps = []
    f = FpgaFinder(args.pcie_address.lower() if args.pcie_address else None)
    devs = f.enum()
    if not devs:
        print('no FPGA found')
        sys.exit(1)

    for d in devs:
        print('sbdf: {segment:04x}:{bus:02x}:{dev:02x}.{func:x}'.format(**d))
        print('FPGA dev:', d)
        args.hssi_grps += f.find_hssi_group(d['pcie_address'])
    print("args.hssi_grps{}".format(args.hssi_grps))
    if len(args.hssi_grps) == 0:
        print("Failed to find HSSI feature")
        sys.exit(1)
    if len(args.hssi_grps) > 1:
        print('{} FPGAs are found: {}\nPlease choose one FPGA'
            .format(len(args.hssi_grps), [d[2] for d in args.hssi_grps]))
        sys.exit(1)

    print("fpga uio dev:{}".format(args.hssi_grps[0][0]))

    lp = FPGAHSSILPBK(args)
    if not lp.hssi_loopback_start():
        print("Failed to enable/disable loopback")
        sys.exit(1)

    if args.loopback == 'enable':
        print("hssi loopback enabled to port{}".format(args.port))
    else:
        print("hssi loopback disabled to port{}".format(args.port))


if __name__ == "__main__":
    main()
